packages/IntegrationTest/windows/integrationtest/TestHostHarness.cpp

```diff
@@ -0,0 +1,184 @@
+
+  switch (listenResult) {
+    case ListenResult::Success:
+      co_return;
+
+    case ListenResult::AddressInUse:
+      ShowJSError("Test harness address is already in use. Is there another instance running?");
+      co_return;
+
+    case ListenResult::OtherError:
+      ShowJSError("The test harness ran into an unexpected error trying to listen for commands.");
+      co_return;
+
+    default:
+      ShowJSError("Unexpected ListenResult from TestCommandListener.");
+      co_return;
+  }
+}
+
+void TestHostHarness::OnTestCommand(const TestCommand &command, TestCommandResponse &&response) noexcept {
+  m_pendingResponse = std::move(response);
+  m_commandStartTime = std::chrono::steady_clock::now();
+
+  switch (command.Id) {
+    case TestCommandId::RunTestComponent: {
+      auto componentName = command.payload.GetString();
+      m_rootView.ComponentName(componentName);
+      m_timeoutAction = TimeoutIfNoResponse();
+      break;
+    }
+
+    case TestCommandId::GoToComponent: {
+      auto componentName = command.payload.GetString();
+      m_rootView.ComponentName(componentName);
+      m_pendingResponse->Okay();
+      CompletePendingResponse();
+      break;
+    }
+
+    default: {
+      ShowJSError("Unexpected command ID from test runner");
+      m_pendingResponse->UnknownError("Unexpected command ID from test runner");
+      CompletePendingResponse();
+      break;
+    }
+  }
+}
+
+void TestHostHarness::OnTestCompleted(steady_clock::time_point eventTime) noexcept {
+  OnTestPassed(eventTime, true);
+}
+
+void TestHostHarness::OnTestPassed(steady_clock::time_point eventTime, bool passed) noexcept {
+  // Some tests like AppEventsTest can misbehave and fire many events at once
+  // if reloaded. We don't have corelation IDs, so just try to filter out
+  // events the JS thread saw before we started the current command.
+  if (eventTime < m_commandStartTime) {
+    return;
+  }
+
+  if (m_pendingResponse) {
+    m_pendingResponse->TestPassed(passed);
+    CompletePendingResponse();
+  }
+}
+
+IAsyncAction TestHostHarness::TimeoutIfNoResponse() noexcept {
+  constexpr auto CommandTimeout = 20s;
+
+  auto weakThis = get_weak();
+  co_await CommandTimeout;
+
+  auto cancellationToken = co_await winrt::get_cancellation_token();
+  if (auto strongThis = weakThis.get() && m_pendingResponse && !cancellationToken()) {
+    m_pendingResponse->Timeout();
+    CompletePendingResponse();
+  }
+}
+
+void TestHostHarness::CompletePendingResponse() noexcept {
+  if (m_pendingResponse) {
+    m_pendingResponse.reset();
+
+    if (m_timeoutAction) {
+      m_timeoutAction.Cancel();
+    }
+  }
+}
+
+void TestHostHarness::ShowJSError(std::string_view err) noexcept {
+  m_context.CallJSFunction(L"RCTLog", L"logToConsole", "error", err);
+}
+
+void TestHostHarnessRedboxHandler::ShowNewError(IRedBoxErrorInfo info, RedBoxErrorType /*type*/) {
+  if (auto strongHarness = m_weakHarness.get()) {
+    if (strongHarness->m_pendingResponse) {
```

Comment:
I'm not actually sure if this actually gets called on the UI creating thread/UI thread. Need to dispatch to the UI thread and probably do the same time-based coalescing.